server/src/main/java/org/elasticsearch/action/admin/cluster/migration/PostFeatureUpgradeResponse.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.admin.cluster.migration;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+public class PostFeatureUpgradeResponse extends ActionResponse implements ToXContentObject {
+
+    private final boolean accepted;
+    private final List<Feature> features;
+    private final String reason;
+    private final ElasticsearchException elasticsearchException;
+
+    public PostFeatureUpgradeResponse(boolean accepted, List<Feature> features, String reason, ElasticsearchException exception) {
+        this.accepted = accepted;
+        this.features = features;
+        this.reason = reason;
+        this.elasticsearchException = exception;
+    }
+
+    public PostFeatureUpgradeResponse(StreamInput in) throws IOException {
+        super(in);
+        this.accepted = in.readBoolean();
+        this.features = in.readList(Feature::new);
+        this.reason = in.readOptionalString();
+        this.elasticsearchException = in.readOptionalWriteable(ElasticsearchException::new);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.field("accepted", this.accepted);
+        if (accepted) {
+            builder.startArray("features");
+            for (Feature feature : this.features) {
+                feature.toXContent(builder, params);
+            }
+            builder.endArray();
+        }
+        if (Objects.nonNull(this.reason)) {
+            builder.field("reason", this.reason);
+        }
+        if (Objects.nonNull(this.elasticsearchException)) {
+            builder.field("exception");
+            builder.startObject();
+            elasticsearchException.toXContent(builder, params);
+            builder.endObject();
+        }
+        builder.endObject();
+        return builder;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeBoolean(this.accepted);
+        out.writeList(this.features);
```

Comment:
If `features` can be `null` you'll want to wrap this in a null check similar to [`writeOptionalStringCollection`](https://github.com/elastic/elasticsearch/blob/944b3f3b56caf910ab1730afe1114bd154e59ec6/server/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java#L1244-L1251), otherwise this will explode when `writeList` tries to call `.size()` on the null object.

I'm not sure why we don't have a `writeOptionalList` on `StreamOutput` already. Possibly to encourage using empty lists instead of `null`?

I think the best solution might be to make `features` non-nullable - either throw in the constructor when it's null, or do what's done today in `getFeatures()` and just replace the null with an empty list.

Reply:
I'll convert null to an empty list in the constructor.